### PR TITLE
Core, Test: Refactor tests to support ClickHouse Cloud 

### PR DESCRIPTION
**Pull request by mzitnik**
## Summary
Add Cloud test according to @pan3793 development https://github.com/pan3793/spark-clickhouse-connector/tree/cloud-test

## Checklist
Delete items not relevant to your PR:
- [ ] Unit and integration tests covering the common scenarios were added
- [ ] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

### CODE REVIEW

File: spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/WriteDistributionAndOrderingSuite.scala [hunk 1]
```diff
@@ -78,7 +85,7 @@ class WriteDistributionAndOrderingSuite extends SparkClickHouseSingleTest {
     WRITE_REPARTITION_BY_PARTITION.key -> repartitionByPartition.toString,
     WRITE_LOCAL_SORT_BY_KEY.key -> localSortByKey.toString
   ) {
-    if (!ignoreUnsupportedTransform && repartitionByPartition) {
+    if (!ignoreUnsupportedTransform && repartitionByPartition && !isCloud) {
```
Comment (pan3793) [inline spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/WriteDistributionAndOrderingSuite.scala:88]: why does it not work on clickhouse cloud?

File: clickhouse-core/src/main/scala/com/clickhouse/spark/client/NodeClient.scala [hunk 2]
```diff
@@ -211,4 +213,6 @@ class NodeClient(val nodeSpec: NodeSpec) extends AutoCloseable with Logging {
        |$sql
        |""".stripMargin
   )
+  def ping(timeout: Int = timeout) =
```
Comment (pan3793) [inline clickhouse-core/src/main/scala/com/clickhouse/spark/client/NodeClient.scala:216]: where is the caller?

File: spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala [hunk 3]
```diff
@@ -176,7 +183,7 @@ class ClickHouseDataTypeSuite extends SparkClickHouseSingleTest {
   private def testDataType(valueColDef: String)(prepare: (String, String) => Unit)(validate: DataFrame => Unit)
     : Unit = {
     val db = "test_kv_db"
-    val tbl = "test_kv_tbl"
+    val tbl = s"test_kv_tbl_${valueColDef}".replace("(", "_").replace(")", "_").replace(",", "_")
```
Comment (pan3793) [inline spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala:186]: seems we don't need this change after disabling parallel testing

File: spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseTableDDLSuite.scala [hunk 4]
```diff
@@ -14,15 +14,22 @@
...
+class ClickHouseSingleTableDDLSuite extends ClickHouseTableDDLSuite with ClickHouseSingleMixIn
+
+abstract class ClickHouseTableDDLSuite extends SparkClickHouseSingleTest {
 
   import testImplicits._
 
   test("clickhouse command runner") {
     withTable("default.abc") {
-      runClickHouseSQL("CREATE TABLE default.abc(a UInt8) ENGINE=Log()")
+      runClickHouseSQL("CREATE TABLE default.abc(a UInt8) ENGINE=Memory()")
```
Comment (pan3793) [inline spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseTableDDLSuite.scala:32]: the cloud version does not support log engine?

File: spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/WriteDistributionAndOrderingSuite.scala (hunk 1 above)
Comment (mzitnik) [reply spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/WriteDistributionAndOrderingSuite.scala:88]: It does not throw  the exception on Cloud that we are expecting on the Cloud Service

File: spark-3.4/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala [hunk 5]
```diff
@@ -176,7 +183,7 @@ class ClickHouseDataTypeSuite extends SparkClickHouseSingleTest {
   private def testDataType(valueColDef: String)(prepare: (String, String) => Unit)(validate: DataFrame => Unit)
     : Unit = {
     val db = "test_kv_db"
-    val tbl = "test_kv_tbl"
+    val tbl = s"test_kv_tbl_${valueColDef}".replace("(", "_").replace(")", "_").replace(",", "_")
```
Comment (pan3793) [inline spark-3.4/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala:186]: ditto

File: clickhouse-core/src/main/scala/com/clickhouse/spark/client/NodeClient.scala [hunk 6]
```diff
@@ -53,7 +54,6 @@ class NodeClient(val nodeSpec: NodeSpec) extends AutoCloseable with Logging {
       "Spark-ClickHouse-Connector"
     }
   }
-
```
Comment (pan3793) [inline clickhouse-core/src/main/scala/com/clickhouse/spark/client/NodeClient.scala:56]: unnecessary change, and please keep one empty line between functions/values that ha multi lines defination

File: clickhouse-core/src/testFixtures/scala/com/clickhouse/spark/base/ClickHouseProvider.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        opts.put("ssl", "true")
+      NodeClient(NodeSpec(
+        clickhouseHost,
+        Some(clickhouseHttpPort),
+        Some(clickhouseTcpPort),
+        protocol,
+        username = clickhouseUser,
+        database = clickhouseDatabase,
+        password = clickhousePassword,
+        options = opts
```
Comment (pan3793) [inline clickhouse-core/src/testFixtures/scala/com/clickhouse/spark/base/ClickHouseProvider.scala:49]: `opts` could be inline 
```
import scala.collection.JavaConverters._

options = Map("ssl" -> isSslEnabled.toString).asJava
````

File: spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseTableDDLSuite.scala (hunk 4 above)
Comment (mzitnik) [reply spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseTableDDLSuite.scala:32]: Correct

File: clickhouse-core/src/testFixtures/scala/com/clickhouse/spark/base/ClickHouseProvider.scala [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  def clickhouseUser: String
+  def clickhousePassword: String
+  def clickhouseDatabase: String
+  def clickhouseVersion: ClickHouseVersion
+  def isSslEnabled: Boolean
+  def isCloud: Boolean = false
+
+  def withNodeClient(protocol: ClickHouseProtocol = HTTP)(block: NodeClient => Unit): Unit =
+    Utils.tryWithResource {
+      import scala.collection.JavaConverters._
```
Comment (pan3793) [inline clickhouse-core/src/testFixtures/scala/com/clickhouse/spark/base/ClickHouseProvider.scala:38]: nit: move this to file header

**Comment by pan3793** [comment]
overall lgtm if tests pass, just small nits. thanks for making it work

File: spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala (hunk 3 above)
Comment (mzitnik) [reply spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala]: This actually happens when testing against the docker I will disable also parallel testing and test

**Comment by mzitnik** [comment]
Thanks, @pan3793 yes we have some improvements that we need to make

**Comment by pan3793** [comment]
we'd better fix the flaky issues before merging to keep the CI stable ...

sorry, it's midnight in my timezone, I have to go to bed now, will take another look tomorrow

**Comment by mzitnik** [comment]
When we run it locally, it works perfectly. Maybe somehow related to the Github

File: spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala (hunk 3 above)
Comment (mzitnik) [reply spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseDataTypeSuite.scala]: So, I have run some tests related to old ClickHouse versions. Once I added a sleep there, it worked. I will try to understand why it happens only on those versions of ClickHouse. Maybe for now a conditional sleep will do the trick

**Review by pan3793 — APPROVED** [review]

**Review by mshustov — APPROVED** [review]

**Pull request merged**